nakama/main.go

```diff
@@ -273,46 +273,86 @@ func handleShowPersona(ctx context.Context, logger runtime.Logger, db *sql.DB, n
 // set up RPC wrappers around each one.
 func initCardinalEndpoints(logger runtime.Logger, initializer runtime.Initializer) error {
-	endpoints, err := cardinalListAllEndpoints()
+	//endpoints, err := cardinalListAllEndpoints()
+
+	endpointStruct, err := cardinalGetEndpointsStruct()
 	if err != nil {
-		return fmt.Errorf("failed to get list of cardinal endpoints: %w", err)
+		return err
 	}
 
-	for _, e := range endpoints {
-		logger.Debug("registering: %v", e)
-		currEndpoint := e
-		if currEndpoint[0] == '/' {
-			currEndpoint = currEndpoint[1:]
+	txEndpoints := endpointStruct.TxEndpoints
+	queryEndpoints := endpointStruct.QueryEndpoints
+
+	createSignedPayload := func(payload string, endpoint string, nk runtime.NakamaModule, ctx context.Context) (io.Reader, error) {
+		logger.Debug("The %s endpoint requires a signed payload", endpoint)
+		signedPayload, err := makeSignedPayload(ctx, nk, payload)
+		if err != nil {
+			return nil, err
 		}
-		err := initializer.RegisterRpc(currEndpoint, func(ctx context.Context, logger runtime.Logger, db *sql.DB, nk runtime.NakamaModule, payload string) (string, error) {
-			logger.Debug("Got request for %q", currEndpoint)
+		return signedPayload, nil
+	}
 
-			signedPayload, err := makeSignedPayload(ctx, nk, payload)
-			if err != nil {
-				return logError(logger, "unable to make signed payload: %v", err)
-			}
+	createUnsignedPayload := func(payload string, endpoint string, _ runtime.NakamaModule, _ context.Context) (io.Reader, error) {
+		payloadBytes := []byte(payload)
+		formattedPayloadBuffer := bytes.NewBuffer([]byte{})
+		if !json.Valid(payloadBytes) {
+			return nil, fmt.Errorf("data %q is not valid json", string(payloadBytes))
+		}
+		err = json.Compact(formattedPayloadBuffer, payloadBytes)
+		if err != nil {
+			return nil, err
+		}
+		return formattedPayloadBuffer, nil
+	}
 
-			req, err := http.NewRequestWithContext(ctx, "POST", makeURL(currEndpoint), signedPayload)
-			if err != nil {
-				return logError(logger, "request setup failed for endpoint %q: %v", currEndpoint, err)
+	registerEndpoints := func(endpoints []string, createPayload func(string, string, runtime.NakamaModule, context.Context) (io.Reader, error)) error {
+		for _, e := range endpoints {
+			logger.Debug("registering: %v", e)
+			currEndpoint := e
+			if currEndpoint[0] == '/' {
+				currEndpoint = currEndpoint[1:]
 			}
-			resp, err := http.DefaultClient.Do(req)
+			err := initializer.RegisterRpc(currEndpoint, func(ctx context.Context, logger runtime.Logger, db *sql.DB, nk runtime.NakamaModule, payload string) (string, error) {
+				logger.Debug("Got request for %q", currEndpoint)
+				var resultPayload io.Reader
+				resultPayload, err = createPayload(payload, currEndpoint, nk, ctx)
+				if err != nil {
+					return logError(logger, "unable to make payload: %v", err)
+				}
+
+				req, err := http.NewRequestWithContext(ctx, "POST", makeURL(currEndpoint), resultPayload)
```

Comment:
If your goal is to share the code that actually makes the HTTP request, break this code out into a top level function.

Reply:
No, the goal here was not to make these top level functions. These functions were created in the spirit of closures, only in this context and never anywhere else. The entire point was to re-use this logic twice for each set of tx and query endpoints. imo i think it's fine.

Reply:
@technicallyty @jerargus 

To give more clarity there are two ways to do this logic:

First way:
```
func A(){
   ...
}

func B(){
   ....
}

func M(inputFunc func()) {
   ....stuff
   inputFunc()
   ...stuff
}

M(A)
M(B)
```

Or:
```
func A(){
   ...
}

func B(){
   ....
}

for _, inputFunc :=range []func(){A, B} {
   ....stuff
   inputFunc()
   ...stuff
}
```

I'm doing the former rather then the latter method. The logic within M is never suppose to be exposed anyway.

Reply:
i think its ok as is for now - can break this out if it becomes a problem later